iati/core/rulesets.py

```diff
@@ -1,45 +1,841 @@
+        except TypeError:
+            # a TypeError is raised in python3 if either of the dates is None
+            return None
+        return True
+
+
+class RuleDependent(Rule):
+    """Representation of a Rule that checks that if one of the Elements or Attributes in a given `path` exists then all its dependent Elements or Attributes must also exist."""
+
+    def __init__(self, context, case):
+        """Initialise a `dependent` rule."""
+        self.name = 'dependent'
+
+        super(RuleDependent, self).__init__(context, case)
+
+    def __str__(self):
+        """Return string stating what TestRuleDependent is checking."""
+        if len(self.paths) == 1:
+            return 'Within each `{self.context}`, either `{self.paths[0]}` exists or it does not. As such, this Rule is always True.'.format(**locals())
+        return 'Within each `{self.context}`, either none of `{0}` must exist, or they must all exist.'.format('` or `'.join(self.paths), **locals())
+
+    def _check_against_Rule(self, context_element):
+        """Assert that either all given `paths` or none of the given `paths` exist for the `context_element`.
+
+        Args:
+            context_element (Element): An XML Element.
+
+        Returns:
+            bool: Return `True` when all dependent `paths` are found in the Dataset, if any exist.
+                  Return `False` when only some of the dependent `paths` are found in the Dataset.
+
+        """
+        unique_paths = set(self.paths)
+        found_paths = 0
+        for path in unique_paths:
+            results = context_element.xpath(path)
+            if results != list():
+                found_paths += 1
+
+        if found_paths not in [0, len(unique_paths)]:
+            return False
+        return True
+
+
+class RuleNoMoreThanOne(Rule):
+    """Representation of a Rule that checks that there is no more than one Element or Attribute matching a given XPath."""
+
+    def __init__(self, context, case):
+        """Initialise a `no_more_than_one` rule."""
+        self.name = 'no_more_than_one'
+
+        super(RuleNoMoreThanOne, self).__init__(context, case)
+
+    def __str__(self):
+        """Return string stating what RuleNoMoreThanOne is checking."""
+        if len(self.paths) == 1:
+            return '`{self.paths[0]}` must occur zero or one times within each `{self.context}`.'.format(**locals())
+        return 'There must be no more than one element or attribute matched at `{0}` within each `{self.context}`.'.format('` or `'.join(self.paths), **locals())
+
+    def _check_against_Rule(self, context_element):
+        """Check `context_element` has no more than one result for a specified Element or Attribute.
+
+        Args:
+            context_element (Element): An XML Element.
+
+        Returns:
+            bool: Return `True` when one or fewer results are found in the Dataset.
+                  Return `False` when one or more results are found in the Dataset.
```

Comment:
`one or more results are` -> `more than one result is`

Reply:
fixed